{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "d15iXj0TvreQ"
   },
   "outputs": [],
   "source": [
    "# Imports\n",
    "from Bio import Entrez, SeqIO\n",
    "import pandas as pd\n",
    "import sqlite3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 367
    },
    "colab_type": "code",
    "executionInfo": {
     "elapsed": 699,
     "status": "error",
     "timestamp": 1537040393462,
     "user": {
      "displayName": "Jessica Wu",
      "photoUrl": "//lh6.googleusercontent.com/-zu-cMxmCEoM/AAAAAAAAAAI/AAAAAAAAAB8/fhUj1O3bNnU/s50-c-k-no/photo.jpg",
      "userId": "116220866113085399826"
     },
     "user_tz": 420
    },
    "id": "GB-yyCPPvred",
    "outputId": "4bd7ec2c-a14d-430f-be25-f569a508c64b"
   },
   "outputs": [],
   "source": [
    "# Define Entrez arguments\n",
    "Entrez.email = 'jessica.wu@berkeley.edu'\n",
    "organisms = ['escherichia e. coli', 'drosophila melanogaster fruit fly', 'homo sapiens human']\n",
    "enzyme_dict = {'glycolysis': ['phosphoglucomutase','aldose 1-epimerase', 'phosphofructokinase', 'phosphopyruvate hydratase'],\n",
    "           'citric acid cycle': ['succinate dehydrogenase', 'aconitase', 'dihydrolipoyllysine-residue acetyltransferase', 'citrate synthase'],\n",
    "           'pentose phophate pathway': ['glucose-6-phosphate isomerase','transketolase', 'malate dehydrogenase', 'phosphogluconolactonase']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gene_fields(query):\n",
    "    # Get name, ID, organism, chr, start and end fields from gene database\n",
    "    gene_handle = Entrez.esearch(db='gene', term=query, sort='relevance')\n",
    "    i = Entrez.read(gene_handle)['IdList'][0]    # only fetch most relevant result\n",
    "    xml_handle = Entrez.efetch(db='gene', id=i, rettype='', retmode='xml')\n",
    "    \n",
    "    # Parse XML\n",
    "    records = Entrez.parse(xml_handle)\n",
    "    for record in records:\n",
    "        geneid = record['Entrezgene_track-info']['Gene-track']['Gene-track_geneid']    # gene ID\n",
    "        organism_name = record['Entrezgene_source']['BioSource']['BioSource_org']['Org-ref']['Org-ref_taxname']   # organism name\n",
    "        locus_info = record['Entrezgene_locus'][0]\n",
    "        pos_info = locus_info['Gene-commentary_seqs'][0]['Seq-loc_int']['Seq-interval']\n",
    "        start_pos = pos_info['Seq-interval_from']     # start pos\n",
    "        end_pos = pos_info['Seq-interval_to']         # end pos\n",
    "\n",
    "    return(geneid, organism_name, start_pos, end_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_sequence(enzyme_name, organism_name, start_pos, end_pos):\n",
    "    # Get sequence field from nucleotide database (might not match data from gene database)\n",
    "    query = enzyme_name + ' ' + organism_name\n",
    "    nt_handle = Entrez.esearch(db='nucleotide', term=query, sort='relevance')\n",
    "    i = Entrez.read(nt_handle)['IdList'][0]   # only fetch most relevant result\n",
    "    fasta_handle = Entrez.efetch(db='nucleotide', id=i, rettype='fasta', retmode='text')\n",
    "\n",
    "    # Parse FASTA\n",
    "    for record in SeqIO.parse(fasta_handle, \"fasta\"):\n",
    "        record_str = str(record)\n",
    "        if 'complete' or 'whole' or 'chromosome' in record_str:\n",
    "            sequence = record.seq[start_pos:end_pos]    # this probably gets the wrong region\n",
    "        else:\n",
    "            sequence = record.seq\n",
    "\n",
    "    return(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch query items from Entrez (gene and nucleotide databases) and parse out gene table fields\n",
    "gene_entries = []\n",
    "for organism in organisms:\n",
    "    for pathway, enzyme_list in enzyme_dict.items():\n",
    "        for enzyme in enzyme_list:\n",
    "            query = enzyme + ' ' + organism\n",
    "            geneid, organism_name, start_pos, end_pos = parse_gene_fields(query)\n",
    "            sequence = parse_sequence(enzyme, organism, int(start_pos), int(end_pos))\n",
    "\n",
    "            gene_entry = (enzyme, int(geneid), organism_name, int(start_pos), int(end_pos), str(sequence))\n",
    "            gene_entries.append(gene_entry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             name        id  \\\n",
      "1                              aldose 1-epimerase    944943   \n",
      "2                             phosphofructokinase    948412   \n",
      "3                       phosphopyruvate hydratase   8156768   \n",
      "4                         succinate dehydrogenase    945402   \n",
      "5                                       aconitase    946724   \n",
      "6   dihydrolipoyllysine-residue acetyltransferase   7396899   \n",
      "7                                citrate synthase    945323   \n",
      "8                   glucose-6-phosphate isomerase  13700005   \n",
      "9                                   transketolase    947420   \n",
      "10                           malate dehydrogenase    947854   \n",
      "11                        phosphogluconolactonase    946398   \n",
      "12                             phosphoglucomutase     44010   \n",
      "13                             aldose 1-epimerase     40406   \n",
      "14                            phosphofructokinase     36060   \n",
      "15                      phosphopyruvate hydratase     33351   \n",
      "16                        succinate dehydrogenase     37228   \n",
      "17                                      aconitase     44149   \n",
      "18  dihydrolipoyllysine-residue acetyltransferase     34021   \n",
      "19                               citrate synthase     31579   \n",
      "20                  glucose-6-phosphate isomerase     35886   \n",
      "21                                  transketolase     41027   \n",
      "22                           malate dehydrogenase     47173   \n",
      "23                        phosphogluconolactonase     33783   \n",
      "24                             phosphoglucomutase      5236   \n",
      "25                             aldose 1-epimerase    130589   \n",
      "26                            phosphofructokinase      7157   \n",
      "27                      phosphopyruvate hydratase      2023   \n",
      "28                        succinate dehydrogenase      6390   \n",
      "29                                      aconitase        48   \n",
      "30  dihydrolipoyllysine-residue acetyltransferase      1629   \n",
      "31                               citrate synthase       348   \n",
      "32                  glucose-6-phosphate isomerase      2821   \n",
      "33                                  transketolase      7086   \n",
      "34                           malate dehydrogenase     26227   \n",
      "35                        phosphogluconolactonase      9563   \n",
      "\n",
      "                                     organism      start        end  \\\n",
      "1   Escherichia coli str. K-12 substr. MG1655     787796     788836   \n",
      "2   Escherichia coli str. K-12 substr. MG1655    4107551    4108513   \n",
      "3   Escherichia coli 'BL21-Gold(DE3)pLysS AG'     999373    1000671   \n",
      "4   Escherichia coli str. K-12 substr. MG1655     755906     757672   \n",
      "5   Escherichia coli str. K-12 substr. MG1655    1335830    1338505   \n",
      "6                       Escherichia coli LF82     130679     132571   \n",
      "7   Escherichia coli str. K-12 substr. MG1655     753184     754467   \n",
      "8    Escherichia coli O104:H4 str. 2011C-3493    4904188    4905837   \n",
      "9   Escherichia coli str. K-12 substr. MG1655    3079643    3081634   \n",
      "10  Escherichia coli str. K-12 substr. MG1655    3383329    3384267   \n",
      "11  Escherichia coli str. K-12 substr. MG1655     798585     799580   \n",
      "12                    Drosophila melanogaster   16126017   16128530   \n",
      "13                    Drosophila melanogaster   21636946   21639059   \n",
      "14                    Drosophila melanogaster   10109739   10117456   \n",
      "15                    Drosophila melanogaster    1724767    1729635   \n",
      "16                    Drosophila melanogaster   19391862   19395970   \n",
      "17                    Drosophila melanogaster   21168569   21173184   \n",
      "18                    Drosophila melanogaster    7426865    7430328   \n",
      "19                    Drosophila melanogaster    6350628    6361322   \n",
      "20                    Drosophila melanogaster    8910899    8914068   \n",
      "21                    Drosophila melanogaster    8667867    8673701   \n",
      "22                    Drosophila melanogaster   12713102   12722441   \n",
      "23                    Drosophila melanogaster    5547085    5549613   \n",
      "24                               Homo sapiens   63593275   63660244   \n",
      "25                               Homo sapiens   38665909   38734766   \n",
      "26                               Homo sapiens    7668401    7687549   \n",
      "27                               Homo sapiens    8861001    8878685   \n",
      "28                               Homo sapiens   17018721   17054169   \n",
      "29                               Homo sapiens   32384602   32450835   \n",
      "30                               Homo sapiens  100186921  100249863   \n",
      "31                               Homo sapiens   44905748   44909394   \n",
      "32                               Homo sapiens   34353329   34402412   \n",
      "33                               Homo sapiens   53224706   53256113   \n",
      "34                               Homo sapiens  119711795  119744225   \n",
      "35                               Homo sapiens    9234766    9271336   \n",
      "\n",
      "                                             sequence  \n",
      "1   TGTCGCCATTGTGCGTATCGAGCAACTCTACCCGTTCCCGCATAAA...  \n",
      "2   TGGCGGCATAGTTTCATACTGCATCACCCATTCAATCTTATTCAGA...  \n",
      "3   AGGCTCTTCGGCGTTTTGAGGCGATGGTGGAAGAAGTTGCCCGCCA...  \n",
      "4   GAAGGTTTGCGGAACTATCTAAAACGTTGCAGACAAAGGACAAAGC...  \n",
      "5   GGGTGCCGCCAGAGGCCATGCTGGATGGCGTGAGGCGTTACGCAAG...  \n",
      "6   CGGTGTGGGAGTCGCCGCAGACGATGGTCATGCCCGGCAGGGTGAC...  \n",
      "7   GGCACAGCTTACATTTACAACAGCAAGGCCACTTGAACACTGGGAC...  \n",
      "8   AGGTGCCTGGTAGATAAAGGGTTTGTTTGACATAGCTTCCAGCCTG...  \n",
      "9   GAAATAGCGCTTTGCTGAATGTTCCGGCATAAATTACCCGCTGCGG...  \n",
      "10  ACCGGTCGTTTCGTTATCGGTGGCCCGATGGGTGACTGCGGTCTGA...  \n",
      "11  ACGTACTGCTGGTGCTGTTGCTGATCTTTATGGCGACAGCGCCCAT...  \n",
      "12                                                     \n",
      "13  CCGTTTCAAGGGGAAGGAACGACACCGAGACGCGCCGAGATGTCGA...  \n",
      "14                                                     \n",
      "15  CAGTAACTTTTATACTTTTTATTTCACATCAACTAACAAATCTGAA...  \n",
      "16                                                     \n",
      "17  AATTTTGGCGGGAAACAAGCCGGCGGAACATTGCTGACGCGTATGC...  \n",
      "18  GGAGAATCTCCTTTCAAAACACTGTGATACCTCTGTCAAACATTTC...  \n",
      "19  AATAAATTTTTGTATCCTACATCACTAGCTTTCACATTCCCAAAAA...  \n",
      "20  ACATCGAATTCATAAATTTTTATTTACGGTTTTAAACTACGTGCAA...  \n",
      "21  TAAAAATCAAGTTGTGCTTTTGGTTTTAGGCCAACAAAAGTCTATA...  \n",
      "22                                                     \n",
      "23  ACGCGAGAGATGCGGTTAAAGTTCGTTCAACTGGCAAATGCAGTGA...  \n",
      "24  CTATACTAAGATTTCAAATCAAGTTCATATTAAGACAGCACACCTC...  \n",
      "25  TAATTTAGTGGCGTATGAAACATAGTAGGCTCTCAAGTGTTTGCTA...  \n",
      "26  GTATCATCTTTGTTTCAAAGTTAAACAATAAACTAAGTTCCTCTCA...  \n",
      "27  TTCTCATGGGTCACTGAGGCTTTTTATTTTGAGCACAAAACCACCG...  \n",
      "28  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...  \n",
      "29                                                     \n",
      "30  TTGAAACATCCTGTCATTTAAACAGGGAGTTTGCATGTAGGGGATC...  \n",
      "31                                                     \n",
      "32  GAAAACGTNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...  \n",
      "33                                                     \n",
      "34  TTACTTCTCCACCAAGAACTCTAGCCAGTAAGAAAATCGTTAGTGA...  \n",
      "35  AACCAACCAATAGAAAAATGTAACAGATGTTCAATAAATATCTGTT...  \n"
     ]
    }
   ],
   "source": [
    "# Build gene table\n",
    "conn = sqlite3.connect('genes.db')\n",
    "c = conn.cursor()\n",
    "\n",
    "# Create table\n",
    "c.execute('''CREATE TABLE IF NOT EXISTS genes (name TEXT,\n",
    "                                               id INT,\n",
    "                                               organism TEXT,\n",
    "                                               start INT,\n",
    "                                               end INT,\n",
    "                                               sequence TEXT);''')\n",
    "\n",
    "# Insert rows\n",
    "for gene_entry in gene_entries:\n",
    "    # Insert entry into gene table # INSERT OR REPLACE INTO genes\n",
    "    c.execute('''INSERT INTO genes (name, id, organism, start, end, sequence)\n",
    "                 VALUES (?, ?, ?, ?, ?, ?)''', gene_entry)\n",
    "\n",
    "# View table\n",
    "c.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "tables = c.fetchall()\n",
    "for table in tables:\n",
    "    table_name = table[0]\n",
    "    contents = pd.read_sql_query(\"SELECT * from %s\" % table_name, conn)\n",
    "    print(contents[:][1:])   # ignore index\n",
    "    contents.to_csv(table_name + '.csv', index=False)    # convert to csv\n",
    "\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        name  \\\n",
      "0                 glycolysis   \n",
      "1          citric acid cycle   \n",
      "2  pentose phosphate pathway   \n",
      "\n",
      "                                         description  \n",
      "0  Glycolysis is the metabolic pathway that conve...  \n",
      "1  The citric acid cycle – also known as the tric...  \n",
      "2  The pentose phosphate pathway is a metabolic p...  \n"
     ]
    }
   ],
   "source": [
    "# Build pathway table\n",
    "conn = sqlite3.connect('pathways.db')\n",
    "c = conn.cursor()\n",
    "\n",
    "# Create table\n",
    "c.execute('''CREATE TABLE IF NOT EXISTS pathways (name TEXT,\n",
    "                                                  description TEXT);''')\n",
    "\n",
    "# Insert rows\n",
    "c.execute('''INSERT INTO pathways (name, description)\n",
    "             VALUES ('glycolysis', 'Glycolysis is the metabolic pathway that converts glucose into pyruvate. The free energy released in this process is used to form the high-energy molecules ATP and NADH. Glycolysis is a sequence of ten enzyme-catalyzed reactions');''')\n",
    "c.execute('''INSERT INTO pathways (name, description)\n",
    "             VALUES ('citric acid cycle', 'The citric acid cycle – also known as the tricarboxylic acid cycle (TCA) or the Krebs cycle – is a series of chemical reactions used by all aerobic organisms to release stored energy through the oxidation of acetyl-CoA derived from carbohydrates, fats, and proteins into ATP and carbon dioxide.');''')\n",
    "c.execute('''INSERT INTO pathways (name, description)\n",
    "             VALUES ('pentose phosphate pathway', 'The pentose phosphate pathway is a metabolic pathway parallel to glycolysis. It generates NADPH and pentoses as well as ribose 5-phosphate, the last one a precursor for the synthesis of nucleotides. While it does involve oxidation of glucose, its primary role is anabolic rather than catabolic.');''')\n",
    "\n",
    "# View table\n",
    "c.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "tables = c.fetchall()\n",
    "for table in tables:\n",
    "    table_name = table[0]\n",
    "    contents = pd.read_sql_query(\"SELECT * from %s\" % table_name, conn)\n",
    "    print(contents)\n",
    "    contents.to_csv(table_name + '.csv', index=False)    # convert to csv\n",
    "\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             name                   pathway  \\\n",
      "1                              aldose 1-epimerase                glycolysis   \n",
      "2                             phosphofructokinase                glycolysis   \n",
      "3                       phosphopyruvate hydratase                glycolysis   \n",
      "4                         succinate dehydrogenase         citric acid cycle   \n",
      "5                                       aconitase         citric acid cycle   \n",
      "6   dihydrolipoyllysine-residue acetyltransferase         citric acid cycle   \n",
      "7                                citrate synthase         citric acid cycle   \n",
      "8                   glucose-6-phosphate isomerase  pentose phophate pathway   \n",
      "9                                   transketolase  pentose phophate pathway   \n",
      "10                           malate dehydrogenase  pentose phophate pathway   \n",
      "11                        phosphogluconolactonase  pentose phophate pathway   \n",
      "\n",
      "      EC_num  \n",
      "1    5.1.3.3  \n",
      "2   2.7.1.11  \n",
      "3   4.2.1.11  \n",
      "4   1.2.1.16  \n",
      "5    4.2.1.3  \n",
      "6    1.2.4.2  \n",
      "7    2.3.3.1  \n",
      "8    5.3.1.9  \n",
      "9    2.2.1.1  \n",
      "10  1.1.1.37  \n",
      "11  3.1.1.31  \n"
     ]
    }
   ],
   "source": [
    "# Build enzyme table\n",
    "conn = sqlite3.connect('enzymes.db')\n",
    "c = conn.cursor()\n",
    "\n",
    "# Create table\n",
    "c.execute('''CREATE TABLE IF NOT EXISTS enzymes (name TEXT,\n",
    "                                                 pathway TEXT,\n",
    "                                                 EC_num TEXT);''')\n",
    "\n",
    "# Insert rows\n",
    "for pathway, enzyme_list in enzyme_dict.items():\n",
    "    for enzyme in enzyme_list:\n",
    "        enzyme_entry = [enzyme, pathway]\n",
    "        if 'phosphoglucomutase' in enzyme:\n",
    "            enzyme_entry.append('5.4.2.2')\n",
    "        elif 'aldose 1-epimerase' in enzyme:\n",
    "            enzyme_entry.append('5.1.3.3')\n",
    "        elif 'phosphofructokinase' in enzyme:\n",
    "            enzyme_entry.append('2.7.1.11')\n",
    "        elif 'phosphopyruvate hydratase' in enzyme:\n",
    "            enzyme_entry.append('4.2.1.11')\n",
    "        elif 'succinate dehydrogenase' in enzyme:\n",
    "            enzyme_entry.append('1.2.1.16')\n",
    "        elif 'aconitase' in enzyme:\n",
    "            enzyme_entry.append('4.2.1.3')\n",
    "        elif 'dihydrolipoyllysine-residue acetyltransferase' in enzyme:\n",
    "            enzyme_entry.append('1.2.4.2')\n",
    "        elif 'citrate synthase' in enzyme:\n",
    "            enzyme_entry.append('2.3.3.1')\n",
    "        elif 'glucose-6-phosphate isomerase' in enzyme:\n",
    "            enzyme_entry.append('5.3.1.9')\n",
    "        elif 'transketolase' in enzyme:\n",
    "            enzyme_entry.append('2.2.1.1')\n",
    "        elif 'malate dehydrogenase' in enzyme:\n",
    "            enzyme_entry.append('1.1.1.37')\n",
    "        elif 'phosphogluconolactonase' in enzyme:\n",
    "            enzyme_entry.append('3.1.1.31')\n",
    "\n",
    "        c.execute('''INSERT INTO enzymes (name, pathway, EC_num)\n",
    "                     VALUES (?, ?, ?)''', tuple(enzyme_entry))\n",
    "    \n",
    "# View table\n",
    "c.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "tables = c.fetchall()\n",
    "for table in tables:\n",
    "    table_name = table[0]\n",
    "    contents = pd.read_sql_query(\"SELECT * from %s\" % table_name, conn)\n",
    "    print(contents[:][1:])   # ignore index\n",
    "    contents.to_csv(table_name + '.csv', index=False)    # convert to csv\n",
    "\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "name": "Lab4JW.ipynb",
   "provenance": [],
   "version": "0.3.2"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
